src/test/clojure/xtdb/expression_test.clj

```diff
@@ -315,6 +315,232 @@
+        ;; TODO - this fails because we're passing dates around with MICROSECOND precision
+        ;; (t/is (= #xt/interval-mdn ["P0D" "PT10921H14M30.111111111S"] (date-diff "SECOND" 9)))
+        (t/is (= #xt/interval-mdn ["P0D" "PT10921H14M30.111111S"] (date-diff "SECOND" 6)))
+        (t/is (= #xt/interval-mdn ["P0D" "PT10921H14M30.111S"] (date-diff "SECOND" 3)))
+        (t/is (= #xt/interval-mdn ["P0D" "PT10921H14M30.11S"] (date-diff "SECOND" 2)))))
+    
+    (t/testing "with start and end field"
+        (letfn [(date-diff [start end] (project1 (list 'date-diff 'ldt1 'ldt2 start end) test-doc))]
+          (t/is (= #xt/interval-ym "P15M" (date-diff "YEAR" "MONTH"))) 
+          (t/is (= #xt/interval-mdn ["P455D" "PT1H"] (date-diff "DAY" "HOUR")))
+          (t/is (= #xt/interval-mdn ["P455D" "PT1H14M"] (date-diff "DAY" "MINUTE")))
+          (t/is (= #xt/interval-mdn ["P455D" "PT1H14M30S"] (date-diff "DAY" "SECOND"))) 
+          (t/is (= #xt/interval-mdn ["P0D" "PT10921H14M"] (date-diff "HOUR" "MINUTE")))
+          (t/is (= #xt/interval-mdn ["P0D" "PT10921H14M30S"] (date-diff "HOUR" "SECOND"))) 
+          (t/is (= #xt/interval-mdn ["P0D" "PT10921H14M30S"] (date-diff "MINUTE" "SECOND")))))
+    
+    (t/testing "with start, end field and precision"
+      (letfn [(date-diff [start end precision] (project1 (list 'date-diff 'ldt1 'ldt2 start end precision) test-doc))] 
+        (t/is (= #xt/interval-mdn ["P455D" "PT1H14M30.1S"] (date-diff "DAY" "SECOND" 1)))
+        (t/is (= #xt/interval-mdn ["P0D" "PT10921H14M30.111S"] (date-diff "HOUR" "SECOND" 3))) 
+        (t/is (= #xt/interval-mdn ["P0D" "PT10921H14M30.111111S"] (date-diff "MINUTE" "SECOND" 6)))))))
+
+(t/deftest test-date-diff-zoned-date-time
+  (let [test-doc {:xt$id :foo,
+                  :zdt1 (ZonedDateTime/of 2022 3 12 13 44 52 222222222 (ZoneId/of "America/Coral_Harbour"))
+                  :zdt2 (ZonedDateTime/of 2021 1 11 13 40 26 111111111 (ZoneId/of "America/Cordoba")) }]
+
+    (t/testing "with only start-field"
+      (letfn [(date-diff [start] (project1 (list 'date-diff 'zdt1 'zdt2 start) test-doc))]
+        (t/is (= #xt/interval-ym "P12M" (date-diff "YEAR")))
+        (t/is (= #xt/interval-ym "P14M" (date-diff "MONTH")))
+        (t/is (= #xt/interval-mdn ["P425D" "PT0S"] (date-diff "DAY")))
+        (t/is (= #xt/interval-mdn ["P0D" "PT10202H"] (date-diff "HOUR")))
+        (t/is (= #xt/interval-mdn ["P0D" "PT10202H4M"] (date-diff "MINUTE")))
+        (t/is (= #xt/interval-mdn ["P0D" "PT10202H4M26S"] (date-diff "SECOND")))))
+
+    (t/testing "with start-field and precision"
+      (letfn [(date-diff [start precision] (project1 (list 'date-diff 'zdt1 'zdt2 start precision) test-doc))]
+        ;; TODO - this fails because we're passing dates around with MICROSECOND precision
+        ;; (t/is (= #xt/interval-mdn ["P0D" "PT10921H14M30.111111111S"] (date-diff "SECOND" 9)))
```

Comment:
Worth pointing out - not sure if theres a way to force timestamps that are passed into codegens/represented into nanos? Seems like typically we pass around all our timestamps in `microseconds`. I've similarly passed casted to timestamp-local with a `:micro` unit - if we'd prefer to change this to `nanos` we can do.

Reply:
this isn't quite true - we convert it into the unit of interest. see `with-arg-unit-conversion` and its usages